File: app/services/nutrition_service.py
"""
Nutrition assessment service.

Computes Z-scores using the WHO LMS method and classifies nutritional status.

LMS Method:
  Z = ((measurement / M) ** L - 1) / (L * S)    when L != 0
  Z = ln(measurement / M) / S                     when L == 0

Height-for-Age (HAZ) and Weight-for-Height (WHZ) both use authoritative
WHO Excel LMS parameters.
"""
import math
from typing import Optional

from config import WastingStatus, ZSCORE_CLASSIFICATIONS
from app.services.who_data_service import WHODataService


class NutritionService:
    def __init__(self, who_data: WHODataService):
        self.who_data = who_data

    def compute_haz(
        self, sex: str, age_months: int, height_cm: float
    ) -> Optional[float]:
        """Compute Height-for-Age Z-score using WHO LMS parameters."""
        lms = self.who_data.get_haz_lms(sex, age_months)
        if lms is None or height_cm <= 0:
            return None
        return self._lms_zscore(height_cm, *lms)

    def compute_whz(
        self, sex: str, age_months: float, height_cm: float, weight_kg: float
    ) -> Optional[float]:
        """Compute Weight-for-Height Z-score using LMS method."""
        lms = self.who_data.get_wfh_lms(sex, height_cm, age_months)
        if lms is None:
            return None
        L, M, S = lms
        return self._lms_zscore(weight_kg, L, M, S)

    @staticmethod
    def _lms_zscore(measurement: float, L: float, M: float, S: float) -> float:
        """Compute Z-score from measurement using LMS parameters."""
        if measurement <= 0 or M <= 0 or S <= 0:
            raise ValueError("LMS inputs require positive measurement, M, and S")
        if abs(L) < 1e-6:
            return math.log(measurement / M) / S
        return (((measurement / M) ** L) - 1) / (L * S)

    def classify_haz(self, z: float) -> str:
        """Classify HAZ Z-score into nutritional status."""
        return self._classify(z, ZSCORE_CLASSIFICATIONS["haz"])

    def classify_whz(self, z: float) -> WastingStatus:
        """Classify WHZ Z-score into nutritional status."""
        return self._classify(z, ZSCORE_CLASSIFICATIONS["whz"], WastingStatus.UNKNOWN)

    @staticmethod
    def _classify(z: float, thresholds: dict, unknown="Unknown"):
        for (low, high), label in thresholds.items():
            if low <= z < high:
                return label
        return unknown

File: flutter_app/lib/services/nutrition_service.dart
/// Computes HAZ, WHZ, WAZ and BAZ z-scores using WHO reference data.
///
/// Ported from Python nutrition_service.py — logic is identical.
library;

import 'who_data_service.dart';

class NutritionService {
  final WhoDataService _who;
  NutritionService(this._who);

  /// Compute Height-for-Age Z-score (HAZ).
  ///
  /// Returns null if WHO reference data is unavailable for the given
  /// [sex] and [ageMonths].
  double? computeHaz(String sex, int ageMonths, double heightCm) {
    final lms = _who.getHazLms(sex, ageMonths);
    if (lms == null) return null;
    return WhoDataService.lmsZscore(heightCm, lms.$1, lms.$2, lms.$3);
  }

  /// Compute HAZ using exact decimal age derived from DOB and visit date.
  double? computeHazForAge(String sex, double ageMonths, double heightCm) {
    final lms = _who.getHazLmsForAge(sex, ageMonths);
    if (lms == null) return null;
    return WhoDataService.lmsZscore(heightCm, lms.$1, lms.$2, lms.$3);
  }

  /// Compute Weight-for-Height Z-score (WHZ).
  ///
  /// Uses LMS parameters from [WhoDataService.getWfhLms] and the standard
  /// WHO LMS formula via [WhoDataService.lmsZscore].
  ///
  /// Returns null if LMS data is unavailable for the given [sex] / [heightCm].
  double? computeWhz(
    String sex,
    double ageMonths,
    double heightCm,
    double weightKg,
  ) {
    final lms = _who.getWfhLms(sex, heightCm, ageMonths);
    if (lms == null) return null;
    return WhoDataService.lmsZscore(weightKg, lms.$1, lms.$2, lms.$3);
  }

  /// Compute Weight-for-Age Z-score (WAZ).
  double? computeWaz(String sex, double ageMonths, double weightKg) {
    final lms = _who.getWfaLms(sex, ageMonths);
    if (lms == null) return null;
    return WhoDataService.lmsZscore(weightKg, lms.$1, lms.$2, lms.$3);
  }

  /// Compute BMI-for-Age Z-score (BAZ).
  double? computeBaz(String sex, double ageMonths, double bmi) {
    final lms = _who.getBfaLms(sex, ageMonths);
    if (lms == null) return null;
    return WhoDataService.lmsZscore(bmi, lms.$1, lms.$2, lms.$3);
  }
}
